chore(docs): create `DateRangePicker` and `DateRangePickerField` docs (#646)
commercetools/nimbus: packages/nimbus/src/components/date-range-picker/date-range-picker.dev.mdx

---
title: DateRangePicker Component
tab-title: Implementation
tab-order: 3
---

# DateRangePicker

The DateRangePicker component enables users to select a date range through two flexible input methods: typing dates directly into input fields or selecting from a visual calendar. It's ideal for filtering data, scheduling events, and any scenario requiring start and end date selection.

> 

Reviewer: I'd like to see a `Resources` section somewhere that includes a link to Storybook and other useful links

#### Date Value Types

Three primary date types are available depending on your needs:

```tsx
import { CalendarDate, CalendarDateTime, ZonedDateTime } from '@internationalized/date';

// Date only (no time)
const dateOnly = new CalendarDate(2024, 1, 15);

// Date with time (no timezone)
const dateTime = new CalendarDateTime(2024, 1, 15, 14, 30);

// Date with time and timezone
const zonedDate = new ZonedDateTime(2024, 1, 15, 14, 30, 'America/New_York');
```

#### Creating Date Values

Use helper functions to parse ISO strings or create dates programmatically:

```tsx
import { parseDate, parseDateTime, parseZonedDateTime } from '@internationalized/date';

// Parse ISO date strings
const date = parseDate('2024-01-15');
const dateTime = parseDateTime('2024-01-15T14:30');
const zonedDate = parseZonedDateTime('2024-01-15T14:30[America/New_York]');
```

#### Date Range Structure

The component expects and returns date ranges in this structure:

```tsx
type DateRange = {
  start: CalendarDate | CalendarDateTime | ZonedDateTime;
  end: CalendarDate | CalendarDateTime | ZonedDateTime;
}
```

> **Learn more:** See [Adobe's @internationalized/date documentation](https://react-spectrum.adobe.com/internationalized/date/index.html) for complete API reference and advanced usage.

Reviewer: Could be in a little `tip` section!

        onChange={setValue}
      />
      <Text fontSize="sm" color="neutral.11">
        {value ? `Selected: ${value.start} to ${value.end}` : 'No selection'}
      </Text>
    </Stack>
  );
}
```

Use controlled mode when you need to:
- Synchronize the date range with external state
- Validate or transform date selections
- Clear or programmatically set the date range

### Time Zone Handling

When working with `ZonedDateTime` values, configure time zone display and hour format:

```jsx-live-dev
const App = () => (
  <Stack direction="column" gap="400">
    <DateRangePicker
      granularity="minute"
      hideTimeZone={false}
      hourCycle={24}
    />
    <DateRangePicker
      granularity="minute"
      hideTimeZone={true}
      hourCycle={12}
    />
  </Stack>
)
```

- `hideTimeZone`: Controls visibility of the time zone selector
- `hourCycle`: Set to `12` for 12-hour format with AM/PM, or `24` for 24-hour format

## Component Requirements

### Date Value Types

All date values **must** use `@internationalized/date` types:
- `CalendarDate` for date-only scenarios
- `CalendarDateTime` for dates with times (no timezone)
- `ZonedDateTime` for timezone-aware dates

> ### Accessibility

Author: I see the inclusion of a11y attributes as a component implementation requirement, personally

> For basic usage, the DateRangePicker handles all accessibility requirements internally-**no additional ARIA attributes are required**. However, in specific contexts you should add:

Reviewer: ```suggestion
For basic usage, the DateRangePicker handles all accessibility requirements internally - **no additional ARIA attributes are required**. However, in specific contexts you should add:
```

      />
      <Text fontSize="sm" color="neutral.11">
        {appliedFilter}
      </Text>
    </Stack>
  );
}
```

### Event Scheduling with Times

For scheduling events or appointments, combine date and time selection:

```jsx-live-dev
const App = () => {
  const [eventRange, setEventRange] = useState(null);

  return (
    <Stack direction="column" gap="400">
      <Text fontWeight="500">Event Duration</Text>
      <DateRangePicker
        granularity="minute"
        value={eventRange}
        onChange={setEventRange}
        hourCycle={12}
      />
      {eventRange && (
        <Text fontSize="sm" color="neutral.11">
          Event scheduled from {eventRange.start.toString()} to {eventRange.end.toString()}
        </Text>
      )}
    </Stack>
  );
}
```

## Testing Your Implementation

### Basic Rendering Tests

Verify the component renders with expected elements:

```tsx
import { render, screen } from '@testing-library/react';
import { DateRangePicker } from '@commercetools/nimbus';

describe('DateRangePicker', () => {
  it('renders start and end date inputs', () => {
    render(<DateRangePicker />);

    // Each date input has 3 segments (month, day, year)

>     const dateInputs = screen.getAllByRole('spinbutton');

Reviewer: Is it odd to check for the presence of the date segment? I had to open the documentation site to double check that this is a real thing (it is, of course).

    expect(dateInputs).toHaveLength(6);
  });

  it('renders calendar and clear buttons', () => {
    render(<DateRangePicker />);

    expect(screen.getByRole('button', { name: /open calendar/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /clear/i })).toBeInTheDocument();
  });
});
```

> ### Interaction Tests

Author: What I was trying to go for here was bridging the gap that exists in UI-Kit where you consume a component, but then have to hunt and peck the source to find examples of how to test it when you write unit tests. Given that, do you think this is still worth removing, or do you think there's a better way?

)
```

Use `size="sm"` for compact layouts or dense data tables. Use `size="md"` (default) for standard forms and prominent date selection.

### Visual variants

Choose between solid and ghost variants to match your design context:

```jsx-live-dev
const App = () => (
  <Stack direction="column" gap="400">
    <DateRangePicker variant="solid" />
    <DateRangePicker variant="ghost" />
  </Stack>
)
```

The `solid` variant (default) provides clear visual boundaries. The `ghost` variant offers a minimal appearance suitable for embedded contexts.

### Granularity options

The `granularity` prop controls both the level of date precision and the component's behavior:

```jsx-live-dev
const App = () => (
  <Stack direction="column" gap="400">
    <DateRangePicker granularity="day" />
    <DateRangePicker granularity="minute" />
  </Stack>
)
```

**Behavioral differences:**
- `granularity="day"` (default): Calendar closes automatically after selecting the end date. Time inputs are hidden.
- `granularity="minute"` or other time values: Calendar remains open after date selection, displaying time input fields in the footer. This allows users to set both dates and times in one interaction.

> Available granularity options: `"day"`, `"hour"`, `"minute"`, `"second"`.

Reviewer 2: ```suggestion
- `granularity="day"` (default): Calendar closes automatically after selecting the end date. Time inputs are hidden.
- `granularity="hour" | "minute" | "second"`: Calendar remains open after date selection, displaying time input fields in the footer. This allows users to set both dates and times in one interaction.
```

> ### Controlled state

Reviewer 2: We should also consider having an `uncontrolled` section that shows how to get the selected value from the component

        onChange={setValue}
      />
      <Text fontSize="sm">
        {value ? `Selected: ${value.start} to ${value.end}` : 'No selection'}
      </Text>
    </Stack>
  );
}
```

Use controlled mode when you need to:
- Synchronize the date range with external state
- Validate or transform date selections
- Clear or programmatically set the date range

### Time zone handling

When working with `ZonedDateTime` values, configure time zone display and hour format:

```jsx-live-dev
const App = () => (
  <Stack direction="column" gap="400">
    <DateRangePicker
      granularity="minute"
      hideTimeZone={false}
      hourCycle={24}
    />
    <DateRangePicker
      granularity="minute"
      hideTimeZone={true}
      hourCycle={12}
    />
  </Stack>
)
```

- `hideTimeZone`: Controls visibility of the time zone selector
- `hourCycle`: Set to `12` for 12-hour format with AM/PM, or `24` for 24-hour format

## Component requirements

### Date value types

All date values **must** use `@internationalized/date` types:
- `CalendarDate` for date-only scenarios
- `CalendarDateTime` for dates with times (no timezone)
- `ZonedDateTime` for timezone-aware dates

### Accessibility

> The DateRangePicker handles all accessibility requirements internally. However, for proper tracking and analytics, **always add an `aria-label`** if your use case warrants it:

Reviewer 2: I'm not sure I like the suggestion that `aria-label` is primarily for tracking and analytics. I realize this is wordy, but it is correct for `a11y`, `intl`, and analytics implementation:

```suggestion
The DateRangePicker handles most accessibility requirements internally. However, you must always associate an internationalized label with the component. Visual labels are preferable, and can be set by:
- Using the `DateRangePickerField` (link goes here)
- Associating a `<label>` element with the `DateRangePicker` using `aria-labelledby`:
\```tsx
<label 
  id="label-id"
>
  {intl.formatMessage(labelMessage)}
/>
<DateRangePicker aria-labelledby="label-id" />
\```
- Associating a `<label>` element with the `DateRangePicker` using`htmlFor`:
\```tsx
<label 
  htmlFor="date-range-picker-id"
>
  {intl.formatMessage(labelMessage)}
/>
<DateRangePicker id="date-range-picker-id" />
\```

If your design requires that the label should not be visible, the label should be set using the `aria-label` prop:
\```tsx
<DateRangePicker aria-label={intl.formatMessage(labelMessage)} />
\```

If your use case requires tracking and analytics for this component, it is good practice to add a **persistent**, **unique** id to the component:
\```tsx
const PERSISTENT_ID = "example-date-range-picker"
export const Example = () => (
  <DateRangePicker id={PERSISTENT_ID} />
)
\```
```

Reviewer: I like this a lot more

```tsx
<DateRangePicker aria-label="Report date range" />
```

Using a `<label>` element or `aria-label` is also appropriate when the picker's purpose isn't clear from surrounding content:

```tsx
<label htmlFor="report-dates">Report date range</label>
<DateRangePicker id="report-dates" />

> ```

Reviewer 2: either a label or aria-label is required, this is why in my suggestion above it says that visual labels are preferable

    render(<DateRangePicker onChange={handleChange} />);

    // Interact with date picker...
    // After selection:
    expect(handleChange).toHaveBeenCalledWith(
      expect.objectContaining({
        start: expect.any(CalendarDate),
        end: expect.any(CalendarDate),
      })
    );
  });
});
```

### Testing time selection

When using time granularity, test time input behavior:

```tsx
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { DateRangePicker } from '@commercetools/nimbus';

describe('DateRangePicker with time', () => {
  it('shows time inputs when granularity includes time', async () => {
    const user = userEvent.setup();
    render(<DateRangePicker granularity="minute" />);

    // Open calendar
    const calendarButton = screen.getByRole('button', { name: /open calendar/i });
    await user.click(calendarButton);

    // Time inputs should appear in the popover footer
    await waitFor(() => {
      expect(screen.getByText(/start time/i)).toBeInTheDocument();
      expect(screen.getByText(/end time/i)).toBeInTheDocument();
    });
  });
});
```

## API reference

> <PropsTable id="DateRangePicker" />

Reviewer 2: I think the Props table should come before the testing section. My care amount here is relatively high.

---
title: DateRangePicker Component
tab-title: Implementation
tab-order: 3
---

> # DateRangePicker

Reviewer: I think we need to agree across the board on how we reference component names in the docs because we are _not consistent_

The DateRangePicker component enables users to select a date range through two flexible input methods: typing dates directly into input fields or selecting from a visual calendar. It's ideal for filtering data, scheduling events, and any scenario requiring start and end date selection.

## Getting started

### Import

```jsx
import { DateRangePicker } from '@commercetools/nimbus';
```

For TypeScript projects:
```tsx
import { DateRangePicker, type DateRangePickerProps } from '@commercetools/nimbus';

> ```

Reviewer 3: nit to think about. maybe we can tighten this part up a bit by doing:

```tsx
// Basic import
import { DateRangePicker } from '@commercetools/nimbus';

// TypeScript
import { DateRangePicker, type DateRangePickerProps } from '@commercetools/nimbus';
```

<img width="531" height="248" alt="Screenshot 2025-11-04 at 14 54 57" src="https://github.com/user-attachments/assets/184eaa20-587b-4fa3-bad2-7a435696d640" />

// Date with time and timezone
const zonedDate = new ZonedDateTime(2024, 1, 15, 14, 30, 'America/New_York');
```

### Creating date values

Use helper functions to parse ISO strings or create dates programmatically:

```tsx
import { parseDate, parseDateTime, parseZonedDateTime } from '@internationalized/date';

// Parse ISO date strings
const date = parseDate('2024-01-15');
const dateTime = parseDateTime('2024-01-15T14:30');
const zonedDate = parseZonedDateTime('2024-01-15T14:30[America/New_York]');
```

### Date range structure

The component expects and returns date ranges in this structure:

```tsx
type DateRange = {
  start: CalendarDate | CalendarDateTime | ZonedDateTime;
  end: CalendarDate | CalendarDateTime | ZonedDateTime;
}
```

> [!TIP]\
> See [Adobe's @internationalized/date documentation](https://react-spectrum.adobe.com/internationalized/date/index.html) for complete API reference and advanced usage.

## Usage examples

### Size options

Two size variants are available to match your interface density:

```jsx-live-dev
const App = () => (
  <Stack direction="column" gap="400">
    <DateRangePicker size="sm" />
    <DateRangePicker size="md" />
  </Stack>
)
```

> Use `size="sm"` for compact layouts or dense data tables. Use `size="md"` (default) for standard forms and prominent date selection.

Reviewer 3: i guess we could even just bold the words medium & small in the descriptions and get rid of the mini size headers, too

> Use `size="sm"` for compact layouts or dense data tables. Use `size="md"` (default) for standard forms and prominent date selection.

Reviewer 3: i get that. it was just an idea so you could grab all the info at once instead of reading about the sizes, viewing the options, _then_ reading about the differences. nothing wrong with it, im just a lazy reader. 

if we go this route, will we want to adjust the other one down [below](https://github.com/commercetools/nimbus/pull/646/files#r2492121943) to match this format?

> Use `size="sm"` for compact layouts or dense data tables. Use `size="md"` (default) for standard forms and prominent date selection.

Reviewer: Do we think that engineers will read the design documentation?

> Use `size="sm"` for compact layouts or dense data tables. Use `size="md"` (default) for standard forms and prominent date selection.

Author: @stephsprinkle Not sure, they might. It is the first thing they land on

> Use `size="sm"` for compact layouts or dense data tables. Use `size="md"` (default) for standard forms and prominent date selection.

Author: @valoriecarli Which other one needs to be adjusted? I don't follow

> Use `size="sm"` for compact layouts or dense data tables. Use `size="md"` (default) for standard forms and prominent date selection.

Reviewer 3: 🤔 honestly don't know, but probably only if there is a discrepancy and there is nobody around to ask.

with the exception of an extra header that isn't needed, i only rearranged the existing info (which tbh id like to know as a dev) so the info flows a little more smoothly. but that is a me thing. @jaikamat summed it up that i have enough to math, gimmie what i need to know. 

i do see where it can become repetitive with the guidelines. oh... so many options.
